url5583f3dbb2_XOR19_XScript_tgz-pJ8-xscript_interactive_MainJ8/src/xscript/XExec.java

```diff
@@ -0,0 +1,1927 @@
+				exc.setRaw(rt, "_stackTrace", stackTrace);
+				exception = exc;
+			} catch (Throwable e) {
+				StringWriter sw = new StringWriter();
+				e.printStackTrace(new PrintWriter(sw));
+				String message = sw.toString();
+				XValue value = rt.getModule("sys");
+				XValue type = value.getRaw(rt, "NativeError");
+				XValue exc = rt.alloc(type);
+				XValue m = rt.alloc(message);
+				XValue stackTrace = XUtils.convertStackTrace(rt, getStackTrace(false));
+				exc.setRaw(rt, "_message", m);
+				exc.setRaw(rt, "_cause", XValueNull.NULL);
+				exc.setRaw(rt, "_stackTrace", stackTrace);
+				exception = exc;
+			}
+		}
+		return instrs;
+	}
+
+	private void exec() {
+		int inst = callFrame.readUByte();
+		XOpcode opcode = XOpcode.values()[inst];
+		XValue v1, v2, v3, v4;
+		int i1;
+		String s;
+		if (XFlags.DEBUG) {
+			System.out.println("exec:" + callFrame.getMethodName() + ":" + opcode + ":" + Arrays.toString(Arrays.copyOf(stack, stackpointer)));
+		}
+		switch (opcode) {
+		case ADD:
+			v1 = pop();
+			v2 = pop();
+			if (v1.isNumber() && v2.isNumber()) {
+				if (v1.isFloat() || v2.isFloat()) {
+					v3 = new XValueFloat(v2.getFloat() + v1.getFloat());
+				} else {
+					v3 = XValueInt.valueOf(v2.getInt() + v1.getInt());
+				}
+				push(v3);
+			} else {
+				callBiOperator("+", "__add__", "__add2__", v2, v1);
+			}
+			break;
+		case AND:
+			v1 = pop();
+			v2 = pop();
+			if ((v1.isInt() || v1.isBool()) && (v2.isInt() || v2.isBool())) {
+				if (v1.isBool() && v2.isBool()) {
+					v3 = XValueBool.valueOf(v2.getBool() && v1.getBool());
+				} else {
+					v3 = XValueInt.valueOf(v2.getInt() & v1.getInt());
+				}
+				push(v3);
+			} else {
+				callBiOperator("&", "__and__", "__and2__", v2, v1);
+			}
+			break;
+		case CALL: {
+			i1 = callFrame.readUByte();
+			XValue[] params = new XValue[i1];
+			for (int i = i1 - 1; i >= 0; i--) {
+				params[i] = pop();
+			}
+			v2 = pop();
+			v1 = pop();
+			call(v2, v1, params, null, null, null);
+			break;
+		}
+		case CALL_KW: {
+			i1 = callFrame.readUByte();
+			XValue[] params = new XValue[i1];
+			for (int i = i1 - 1; i >= 0; i--) {
+				params[i] = pop();
+			}
+			String[] kws = new String[i1];
+			int i2 = callFrame.readUByte();
+			for (int i = i2; i < i1; i++) {
+				kws[i] = callFrame.readStringP();
+			}
+			v2 = pop();
+			v1 = pop();
+			call(v2, v1, params, kws, null, null);
+			break;
+		}
+		case CALL_LIST: {
+			i1 = callFrame.readUByte() - 1;
+			XValue[] params = new XValue[i1];
+			v3 = pop();
+			for (int i = i1 - 1; i >= 0; i--) {
+				params[i] = pop();
+			}
+			v2 = pop();
+			v1 = pop();
+			call(v2, v1, params, null, v3, null);
+			break;
+		}
+		case CALL_LIST_KW: {
+			i1 = callFrame.readUByte() - 1;
+			int i2 = callFrame.readUByte();
+			v3 = null;
+			XValue[] params = new XValue[i1];
+			for (int i = i1, j = i1 - 1; i >= 0; i--, j--) {
+				if (i == i2) {
+					v3 = pop();
+					i--;
+					continue;
+				}
+				params[j] = pop();
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 630:637, 708:715.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.